File: crates/gust-core/src/middleware/cors.rs
//! CORS (Cross-Origin Resource Sharing) middleware
//!
//! Handles preflight requests and adds CORS headers.

use crate::{Request, Response, ResponseBuilder, StatusCode, Method};
use super::Middleware;
use smallvec::SmallVec;

/// CORS configuration
#[derive(Clone)]
pub struct CorsConfig {
    /// Allowed origins (empty = all)
    pub origins: SmallVec<[String; 4]>,
    /// Allowed methods
    pub methods: SmallVec<[Method; 8]>,
    /// Allowed headers
    pub headers: SmallVec<[String; 8]>,
    /// Exposed headers
    pub expose_headers: SmallVec<[String; 4]>,
    /// Allow credentials
    pub credentials: bool,
    /// Max age (seconds)
    pub max_age: u32,
}

impl Default for CorsConfig {
    fn default() -> Self {
        Self {
            origins: SmallVec::new(), // Empty = allow all
            methods: smallvec::smallvec![
                Method::Get,
                Method::Post,
                Method::Put,
                Method::Delete,
                Method::Patch,
                Method::Head,
                Method::Options,
            ],
            headers: smallvec::smallvec![
                "Content-Type".to_string(),
                "Authorization".to_string(),
                "X-Requested-With".to_string(),
            ],
            expose_headers: SmallVec::new(),
            credentials: false,
            max_age: 86400, // 24 hours
        }
    }
}

impl CorsConfig {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn allow_origin(mut self, origin: impl Into<String>) -> Self {
        self.origins.push(origin.into());
        self
    }

    pub fn allow_all_origins(mut self) -> Self {
        self.origins.clear();
        self
    }

    pub fn allow_method(mut self, method: Method) -> Self {
        if !self.methods.contains(&method) {
            self.methods.push(method);
        }
        self
    }

    pub fn allow_header(mut self, header: impl Into<String>) -> Self {
        self.headers.push(header.into());
        self
    }

    pub fn expose_header(mut self, header: impl Into<String>) -> Self {
        self.expose_headers.push(header.into());
        self
    }

    pub fn allow_credentials(mut self) -> Self {
        self.credentials = true;
        self
    }

    pub fn max_age(mut self, seconds: u32) -> Self {
        self.max_age = seconds;
        self
    }
}

/// CORS middleware
pub struct Cors {
    config: CorsConfig,
}

impl Cors {
    pub fn new(config: CorsConfig) -> Self {
        Self { config }
    }

    /// Simple CORS - allow all origins
    pub fn permissive() -> Self {
        Self::new(CorsConfig::default().allow_all_origins())
    }

    fn is_origin_allowed(&self, origin: &str) -> bool {
        if self.config.origins.is_empty() {
            return true; // Allow all
        }
        self.config.origins.iter().any(|o| o == origin || o == "*")
    }

    fn methods_string(&self) -> String {
        self.config
            .methods
            .iter()
            .map(|m| m.as_str())
            .collect::<Vec<_>>()
            .join(", ")
    }

    fn headers_string(&self) -> String {
        self.config.headers.join(", ")
    }

    fn add_cors_headers(&self, res: &mut Response, origin: &str) {
        // Origin
        let origin_value = if self.config.origins.is_empty() {
            "*".to_string()
        } else {
            origin.to_string()
        };
        res.headers.push(("Access-Control-Allow-Origin".to_string(), origin_value));

        // Methods
        res.headers.push((
            "Access-Control-Allow-Methods".to_string(),
            self.methods_string(),
        ));

        // Headers
        if !self.config.headers.is_empty() {
            res.headers.push((
                "Access-Control-Allow-Headers".to_string(),
                self.headers_string(),
            ));
        }

        // Expose headers
        if !self.config.expose_headers.is_empty() {
            res.headers.push((
                "Access-Control-Expose-Headers".to_string(),
                self.config.expose_headers.join(", "),
            ));
        }

        // Credentials
        if self.config.credentials {
            res.headers.push((
                "Access-Control-Allow-Credentials".to_string(),
                "true".to_string(),
            ));
        }

        // Max age
        res.headers.push((
            "Access-Control-Max-Age".to_string(),
            self.config.max_age.to_string(),
        ));
    }
}

impl Middleware for Cors {
    fn before(&self, req: &mut Request) -> Option<Response> {
        let origin = req.header("origin").unwrap_or("");

        // No origin header = same-origin request, skip CORS
        if origin.is_empty() {
            return None;
        }

        // Check if origin is allowed
        if !self.is_origin_allowed(origin) {
            return Some(
                ResponseBuilder::new(StatusCode::FORBIDDEN)
                    .body("CORS: Origin not allowed")
                    .build(),
            );
        }

        // Handle preflight (OPTIONS)
        if req.method == Method::Options {
            let mut res = ResponseBuilder::new(StatusCode::NO_CONTENT)
                .body("")
                .build();
            self.add_cors_headers(&mut res, origin);
            return Some(res);
        }

        // Store origin for after() - use extension or similar
        // For now we'll re-extract in after()
        None
    }

    fn after(&self, req: &Request, res: &mut Response) {
        let origin = req.header("origin").unwrap_or("");
        if !origin.is_empty() && self.is_origin_allowed(origin) {
            self.add_cors_headers(res, origin);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cors_permissive() {
        let cors = Cors::permissive();
        assert!(cors.is_origin_allowed("https://example.com"));
        assert!(cors.is_origin_allowed("http://localhost:3000"));
    }

    #[test]
    fn test_cors_specific_origin() {
        let config = CorsConfig::new()
            .allow_origin("https://example.com");
        let cors = Cors::new(config);

        assert!(cors.is_origin_allowed("https://example.com"));
        assert!(!cors.is_origin_allowed("https://other.com"));
    }

    #[test]
    fn test_cors_methods() {
        let config = CorsConfig::new();
        let cors = Cors::new(config);

        assert!(cors.methods_string().contains("GET"));
        assert!(cors.methods_string().contains("POST"));
    }
}

File: packages/serve/src/cors.ts
/**
 * CORS - Cross-Origin Resource Sharing support
 * Wrapper for handling CORS preflight and response headers
 */

import type { Handler, ServerResponse, Wrapper } from '@sylphx/gust-core'
import { response } from '@sylphx/gust-core'
import type { BaseContext } from './context'

export type CorsOptions = {
	/** Allowed origins (string, array, or function) */
	readonly origin?: string | string[] | ((origin: string) => boolean)
	/** Allowed HTTP methods */
	readonly methods?: string[]
	/** Allowed headers */
	readonly allowedHeaders?: string[]
	/** Exposed headers */
	readonly exposedHeaders?: string[]
	/** Allow credentials */
	readonly credentials?: boolean
	/** Max age for preflight cache (seconds) */
	readonly maxAge?: number
	/** Handle preflight automatically */
	readonly preflight?: boolean
}

// Default CORS options
const DEFAULT_METHODS = ['GET', 'HEAD', 'PUT', 'PATCH', 'POST', 'DELETE']
const DEFAULT_ALLOWED_HEADERS = ['Content-Type', 'Authorization', 'X-Requested-With']

/**
 * Check if origin is allowed
 */
const isOriginAllowed = (
	origin: string,
	allowed: string | string[] | ((origin: string) => boolean) | undefined
): boolean => {
	if (!allowed) return true
	if (allowed === '*') return true
	if (typeof allowed === 'string') return origin === allowed
	if (Array.isArray(allowed)) return allowed.includes(origin)
	if (typeof allowed === 'function') return allowed(origin)
	return false
}

/**
 * Get allowed origin for response header
 */
const getAllowedOrigin = (
	origin: string,
	allowed: string | string[] | ((origin: string) => boolean) | undefined
): string => {
	if (!allowed) return '*'
	if (allowed === '*') return '*'
	if (isOriginAllowed(origin, allowed)) return origin
	return ''
}

/**
 * Create CORS headers for response
 */
const createCorsHeaders = (origin: string, options: CorsOptions): Record<string, string> => {
	const headers: Record<string, string> = {}

	// Access-Control-Allow-Origin
	const allowedOrigin = getAllowedOrigin(origin, options.origin)
	if (allowedOrigin) {
		headers['access-control-allow-origin'] = allowedOrigin
	}

	// Access-Control-Allow-Credentials
	if (options.credentials) {
		headers['access-control-allow-credentials'] = 'true'
	}

	// Access-Control-Expose-Headers
	if (options.exposedHeaders && options.exposedHeaders.length > 0) {
		headers['access-control-expose-headers'] = options.exposedHeaders.join(', ')
	}

	// Vary header (important for caching)
	if (allowedOrigin && allowedOrigin !== '*') {
		headers.vary = 'Origin'
	}

	return headers
}

/**
 * Create CORS preflight headers
 */
const createPreflightHeaders = (origin: string, options: CorsOptions): Record<string, string> => {
	const headers = createCorsHeaders(origin, options)

	// Access-Control-Allow-Methods
	const methods = options.methods || DEFAULT_METHODS
	headers['access-control-allow-methods'] = methods.join(', ')

	// Access-Control-Allow-Headers
	const allowedHeaders = options.allowedHeaders || DEFAULT_ALLOWED_HEADERS
	headers['access-control-allow-headers'] = allowedHeaders.join(', ')

	// Access-Control-Max-Age
	if (options.maxAge !== undefined) {
		headers['access-control-max-age'] = options.maxAge.toString()
	}

	return headers
}

/**
 * Create CORS wrapper
 * Handles preflight OPTIONS requests and adds CORS headers to responses
 *
 * Works as both global middleware (BaseContext) and route middleware (Context)
 */
export const cors = <Ctx extends BaseContext = BaseContext>(
	options: CorsOptions = {}
): Wrapper<Ctx> => {
	const handlePreflight = options.preflight !== false

	return (handler: Handler<Ctx>): Handler<Ctx> => {
		return async (ctx: Ctx): Promise<ServerResponse> => {
			const origin = ctx.headers.origin || ''

			// Handle preflight OPTIONS request
			if (handlePreflight && ctx.method === 'OPTIONS') {
				const requestMethod = ctx.headers['access-control-request-method']

				// If it's a CORS preflight request
				if (requestMethod) {
					const preflightHeaders = createPreflightHeaders(origin, options)
					return response(null, {
						status: 204,
						headers: preflightHeaders,
					})
				}
			}

			// Execute the handler
			const res = await handler(ctx)

			// Add CORS headers to response
			const corsHeaders = createCorsHeaders(origin, options)
			const mergedHeaders = { ...res.headers, ...corsHeaders }

			return {
				...res,
				headers: mergedHeaders,
			}
		}
	}
}

/**
 * Simple CORS - allows all origins
 * Convenience wrapper for development
 */
export const simpleCors = <Ctx extends BaseContext = BaseContext>(): Wrapper<Ctx> =>
	cors<Ctx>({
		origin: '*',
		methods: DEFAULT_METHODS,
		allowedHeaders: DEFAULT_ALLOWED_HEADERS,
	})
